File: src/recommender.py
from typing import List, Dict, Tuple, Optional
from dataclasses import dataclass
import csv
import os


SCORING_MODES = {
    # Baseline: mood is dominant, genre is secondary.
    "mood_priority": {
        "mood_exact": 4.0,
        "mood_similar": 2.0,
        "genre_exact": 1.5,
        "genre_related": 0.75,
    },
    # Alternate strategy: genre is dominant, mood is secondary.
    "genre_priority": {
        "mood_exact": 2.0,
        "mood_similar": 1.0,
        "genre_exact": 4.0,
        "genre_related": 2.0,
    },
}

@dataclass
class Song:
    """
    Represents a song and its attributes.
    Required by tests/test_recommender.py
    """
    id: int
    title: str
    artist: str
    genre: str
    mood: str
    energy: float
    tempo_bpm: float
    valence: float
    danceability: float
    acousticness: float
    popularity: float = 50.0
    release_year: int = 2015
    instrumentalness: float = 0.3
    lyrical_sentiment: float = 0.0
    production_complexity: float = 0.5

@dataclass
class UserProfile:
    """
    Represents a user's taste preferences.
    Required by tests/test_recommender.py
    """
    favorite_genre: str
    favorite_mood: str
    target_energy: float
    likes_acoustic: bool
    target_popularity: Optional[float] = None
    target_decade: Optional[int] = None
    target_instrumentalness: Optional[float] = None
    target_lyrical_sentiment: Optional[float] = None
    target_production_complexity: Optional[float] = None

class Recommender:
    """
    OOP implementation of the recommendation logic.
    Required by tests/test_recommender.py
    """
    def __init__(self, songs: List[Song]):
        self.songs = songs

    def recommend(self, user: UserProfile, k: int = 5) -> List[Song]:
        # TODO: Implement recommendation logic
        return self.songs[:k]

    def explain_recommendation(self, user: UserProfile, song: Song) -> str:
        # TODO: Implement explanation logic
        return "Explanation placeholder"

def load_songs(csv_path: str) -> List[Dict]:
    """Load songs from a CSV file and return as a list of dictionaries."""
    print(f"Loading songs from {csv_path}...")
    
    # Get the absolute path, handling relative paths from src/ directory
    if not os.path.isabs(csv_path):
        # If relative path, compute relative to this file's location
        current_dir = os.path.dirname(os.path.abspath(__file__))
        csv_path = os.path.join(current_dir, "..", csv_path)
    
    songs = []
    try:
        with open(csv_path, 'r', encoding='utf-8') as file:
            reader = csv.DictReader(file)
            for row in reader:
                # Convert numeric fields to appropriate types
                song = {
                    'id': int(row['id']),
                    'title': row['title'],
                    'artist': row['artist'],
                    'genre': row['genre'],
                    'mood': row['mood'],
                    'energy': float(row['energy']),
                    'tempo_bpm': float(row['tempo_bpm']),
                    'valence': float(row['valence']),
                    'danceability': float(row['danceability']),
                    'acousticness': float(row['acousticness']),
                    'popularity': float(row.get('popularity', 50)),
                    'release_year': int(row.get('release_year', 2015)),
                    'instrumentalness': float(row.get('instrumentalness', 0.3)),
                    'lyrical_sentiment': float(row.get('lyrical_sentiment', 0.0)),
                    'production_complexity': float(row.get('production_complexity', 0.5)),
                }
                songs.append(song)
        
    
    except FileNotFoundError:
        print(f"Error: File not found at {csv_path}")
        return []
    except Exception as e:
        print(f"Error loading songs: {e}")
        return []
    
    print(f"Successfully loaded {len(songs)} songs.")
    return songs
    
def score_song(user_prefs: Dict, song: Dict, scoring_mode: str = "mood_priority") -> Tuple[float, str]:
    """Score a song against user preferences and return (score, explanation) tuple."""
    score = 0
    reasons = []
    weights = SCORING_MODES.get(scoring_mode, SCORING_MODES["mood_priority"])
    
    # 1. MOOD MATCH (weighted by strategy mode)
    if song['mood'].lower() == user_prefs['favorite_mood'].lower():
        score += weights["mood_exact"]
        reasons.append(f"Mood match: {song['mood']} ({weights['mood_exact']} pts)")
    elif _is_mood_similar(song['mood'], user_prefs['favorite_mood']):
        score += weights["mood_similar"]
        reasons.append(f"Similar mood: {song['mood']} vs {user_prefs['favorite_mood']} ({weights['mood_similar']} pts)")
    else:
        reasons.append(f"Mood mismatch: {song['mood']} vs {user_prefs['favorite_mood']} ({0} pts)")
    
    # 2. ENERGY PROXIMITY (0-2.5 points)
    energy_diff = abs(song['energy'] - user_prefs['target_energy'])
    if energy_diff <= 0.1:
        score += 2.5
        reasons.append(f"Energy perfect match: {song['energy']:.2f} ({2.5} pts)")
    elif energy_diff <= 0.2:
        score += 1.5
        reasons.append(f"Energy close: {song['energy']:.2f} vs {user_prefs['target_energy']:.2f} ({1.5} pts)")
    elif energy_diff <= 0.3:
        score += 1.0
        reasons.append(f"Energy acceptable: {song['energy']:.2f} vs {user_prefs['target_energy']:.2f} ({1.0} pts)")
    else:
        reasons.append(f"Energy mismatch: {song['energy']:.2f} vs {user_prefs['target_energy']:.2f} ({0} pts)")
    
    # 3. GENRE MATCH (weighted by strategy mode)
    if song['genre'].lower() == user_prefs['favorite_genre'].lower():
        score += weights["genre_exact"]
        reasons.append(f"Genre match: {song['genre']} ({weights['genre_exact']} pts)")
    elif _is_genre_related(song['genre'], user_prefs['favorite_genre']):
        score += weights["genre_related"]
        reasons.append(f"Related genre: {song['genre']} ({weights['genre_related']} pts)")
    else:
        reasons.append(f"Genre different: {song['genre']} ({0} pts)")
    
    # 4. ACOUSTICNESS BONUS (0-1.2 points)
    if user_prefs['likes_acoustic']:
        if song['acousticness'] > 0.7:
            score += 1.2
            reasons.append(f"High acousticness: {song['acousticness']:.2f} ({1.2} pts)")
        elif song['acousticness'] > 0.4:
            score += 0.6
            reasons.append(f"Moderate acousticness: {song['acousticness']:.2f} ({0.6} pts)")
        else:
            reasons.append(f"Low acousticness: {song['acousticness']:.2f} ({0} pts)")
    else:
        reasons.append(f"Acousticness not prioritized ({0} pts)")
    
    # 5. VALENCE ALIGNMENT (0 to +0.5 or -0.25 points)
    if song['valence'] < 0.3:
        score += 0.5
        reasons.append(f"Low valence (moody): {song['valence']:.2f} ({0.5} pts)")
    elif song['valence'] > 0.6:
        score -= 0.25
        reasons.append(f"High valence (happy): {song['valence']:.2f} ({-0.25} pts)")
    else:
        reasons.append(f"Neutral valence: {song['valence']:.2f} ({0} pts)")
    
    # 6. DANCEABILITY (minimal 0.3 pts, usually 0 for moody)
    # Skipped for moody recommendations
    reasons.append(f"Danceability: ignored for moody aesthetic ({0} pts)")

    # 7. POPULARITY ALIGNMENT (0-1.5 points)
    default_popularity = 70.0 if user_prefs.get('favorite_mood', '').lower() in ['happy', 'upbeat', 'euphoric'] else 50.0
    target_popularity = float(user_prefs.get('target_popularity', default_popularity))
    popularity_diff = abs(song['popularity'] - target_popularity)
    if popularity_diff <= 10:
        score += 1.5
        reasons.append(f"Popularity aligned: {song['popularity']:.0f} vs {target_popularity:.0f} ({1.5} pts)")
    elif popularity_diff <= 20:
        score += 1.0
        reasons.append(f"Popularity close: {song['popularity']:.0f} vs {target_popularity:.0f} ({1.0} pts)")
    elif popularity_diff <= 30:
        score += 0.5
        reasons.append(f"Popularity acceptable: {song['popularity']:.0f} vs {target_popularity:.0f} ({0.5} pts)")
    else:
        reasons.append(f"Popularity mismatch: {song['popularity']:.0f} vs {target_popularity:.0f} ({0} pts)")

    # 8. RELEASE DECADE ALIGNMENT (0-1.0 points)
    target_decade = int(user_prefs.get('target_decade', 2010))
    song_decade = (int(song['release_year']) // 10) * 10
    decade_gap = abs(song_decade - target_decade)
    if decade_gap == 0:
        score += 1.0
        reasons.append(f"Same decade: {song_decade}s ({1.0} pts)")
    elif decade_gap == 10:
        score += 0.6
        reasons.append(f"Adjacent decade: {song_decade}s vs {target_decade}s ({0.6} pts)")
    elif decade_gap == 20:
        score += 0.2
        reasons.append(f"Older/newer decade: {song_decade}s vs {target_decade}s ({0.2} pts)")
    else:
        reasons.append(f"Decade far: {song_decade}s vs {target_decade}s ({0} pts)")

    # 9. INSTRUMENTALNESS ALIGNMENT (0-1.0 points, linear)
    default_instrumentalness = 0.65 if user_prefs.get('likes_acoustic') else 0.30
    target_instrumentalness = float(user_prefs.get('target_instrumentalness', default_instrumentalness))
    instrumentalness_diff = abs(song['instrumentalness'] - target_instrumentalness)
    instrumentalness_points = max(0.0, 1.0 - (2.0 * instrumentalness_diff))
    score += instrumentalness_points
    reasons.append(
        f"Instrumentalness alignment: song={song['instrumentalness']:.2f}, target={target_instrumentalness:.2f} ({instrumentalness_points:.2f} pts)"
    )

    # 10. LYRICAL SENTIMENT ALIGNMENT (0-1.0 points, linear)
    mood_target_sentiment = {
        'moody': -0.45,
        'melancholic': -0.50,
        'sad': -0.60,
        'dark': -0.55,
        'happy': 0.65,
        'upbeat': 0.60,
        'euphoric': 0.75,
        'chill': 0.15,
        'relaxed': 0.10,
        'focused': 0.05,
        'intense': -0.15,
        'aggressive': -0.35,
    }
    default_sentiment = mood_target_sentiment.get(user_prefs.get('favorite_mood', '').lower(), 0.0)
    target_lyrical_sentiment = float(user_prefs.get('target_lyrical_sentiment', default_sentiment))
    lyrical_sentiment_diff = abs(song['lyrical_sentiment'] - target_lyrical_sentiment)
    lyrical_sentiment_points = max(0.0, 1.0 - lyrical_sentiment_diff)
    score += lyrical_sentiment_points
    reasons.append(
        f"Lyrical sentiment alignment: song={song['lyrical_sentiment']:.2f}, target={target_lyrical_sentiment:.2f} ({lyrical_sentiment_points:.2f} pts)"
    )

    # 11. PRODUCTION COMPLEXITY ALIGNMENT (0-0.8 points, linear)
    default_complexity = 0.45 if user_prefs.get('likes_acoustic') else 0.70
    target_production_complexity = float(user_prefs.get('target_production_complexity', default_complexity))
    complexity_diff = abs(song['production_complexity'] - target_production_complexity)
    complexity_points = max(0.0, 0.8 - (1.6 * complexity_diff))
    score += complexity_points
    reasons.append(
        f"Production complexity alignment: song={song['production_complexity']:.2f}, target={target_production_complexity:.2f} ({complexity_points:.2f} pts)"
    )
    
    # Compile explanation
    explanation = " | ".join(reasons)
    
    return score, explanation


def _is_mood_similar(mood1: str, mood2: str) -> bool:
    """Return True if two moods are semantically similar."""
    mood_groups = {
        'moody': ['melancholic', 'sad', 'dark', 'introspective'],
        'melancholic': ['moody', 'sad', 'dark', 'introspective'],
        'chill': ['relaxed', 'focused', 'peaceful', 'calm'],
        'relaxed': ['chill', 'peaceful', 'calm'],
        'happy': ['euphoric', 'upbeat', 'energetic'],
        'energetic': ['happy', 'intense', 'euphoric'],
    }
    
    mood1_lower = mood1.lower()
    mood2_lower = mood2.lower()
    
    return mood1_lower in mood_groups.get(mood2_lower, []) or \
           mood2_lower in mood_groups.get(mood1_lower, [])


def _is_genre_related(genre1: str, genre2: str) -> bool:
    """Return True if two genres are related or adjacent categories."""
    genre_groups = {
        'alternative rock': ['indie rock', 'rock', 'indie pop'],
        'indie rock': ['alternative rock', 'rock', 'indie pop'],
        'synth-pop': ['pop', 'electronic', 'new wave'],
        'pop': ['synth-pop', 'indie pop', 'disco'],
        'lofi': ['ambient', 'chill', 'jazz'],
        'ambient': ['lofi', 'chill', 'classical'],
        'jazz': ['lofi', 'funk', 'soul'],
        'funk': ['disco', 'jazz', 'pop'],
    }
    
    genre1_lower = genre1.lower()
    genre2_lower = genre2.lower()
    
    return genre1_lower in genre_groups.get(genre2_lower, []) or \
           genre2_lower in genre_groups.get(genre1_lower, [])


def recommend_songs(
    user_prefs: Dict,
    songs: List[Dict],
    k: int = 5,
    scoring_mode: str = "mood_priority",
) -> List[Tuple[Dict, float, str]]:
    """Score all songs and return top k recommendations sorted by score (highest first)."""
    # Score all songs using list comprehension
    scored_songs = [
        (song, *score_song(user_prefs, song, scoring_mode=scoring_mode))
        for song in songs
    ]

    # Sort by score (descending).
    sorted_songs = sorted(scored_songs, key=lambda x: x[1], reverse=True)

    # Diversity penalty pattern: prioritize distinct artists in top-k.
    # First pass keeps at most one song per artist.
    diverse_results: List[Tuple[Dict, float, str]] = []
    seen_artists = set()
    for item in sorted_songs:
        artist = item[0]["artist"]
        if artist in seen_artists:
            continue
        diverse_results.append(item)
        seen_artists.add(artist)
        if len(diverse_results) == k:
            return diverse_results

    # If there are not enough unique artists in the catalog, backfill by score.
    for item in sorted_songs:
        if item in diverse_results:
            continue
        diverse_results.append(item)
        if len(diverse_results) == k:
            break

    return diverse_results

File: frontend/src/engine/recommender.js
/**
 * Music recommender scoring engine — ported from src/recommender.py.
 * Implements the same weighted point-based scoring system.
 */

const MOOD_SIMILARITY = {
  moody: ["melancholic", "sad", "dark", "introspective"],
  melancholic: ["moody", "sad", "dark", "introspective"],
  chill: ["relaxed", "focused", "peaceful", "calm"],
  relaxed: ["chill", "peaceful", "calm"],
  happy: ["euphoric", "upbeat", "energetic"],
  energetic: ["happy", "intense", "euphoric"],
  intense: ["energetic", "aggressive"],
  euphoric: ["happy", "energetic", "upbeat"],
  upbeat: ["happy", "euphoric", "energetic"],
  focused: ["chill", "relaxed"],
  dreamy: ["chill", "relaxed", "nostalgic"],
  nostalgic: ["melancholic", "dreamy", "moody"],
  aggressive: ["intense", "energetic"],
};

const GENRE_RELATIONS = {
  "alternative rock": ["indie rock", "rock", "indie pop"],
  "indie rock": ["alternative rock", "rock", "indie pop"],
  rock: ["alternative rock", "indie rock", "metal"],
  "synth-pop": ["pop", "electronic", "new wave", "synthwave"],
  synthwave: ["synth-pop", "electronic", "pop"],
  pop: ["synth-pop", "indie pop", "disco"],
  "indie pop": ["pop", "indie rock", "alternative rock"],
  lofi: ["ambient", "chill", "jazz"],
  ambient: ["lofi", "chill", "classical"],
  jazz: ["lofi", "funk", "soul"],
  funk: ["disco", "jazz", "pop"],
  disco: ["funk", "pop", "dance"],
  classical: ["ambient", "jazz"],
  metal: ["rock", "alternative rock"],
  reggae: ["funk", "pop"],
};

const SCORING_MODES = {
  mood_priority: {
    mood_exact: 4.0,
    mood_similar: 2.0,
    genre_exact: 1.5,
    genre_related: 0.75,
  },
  genre_priority: {
    mood_exact: 2.0,
    mood_similar: 1.0,
    genre_exact: 4.0,
    genre_related: 2.0,
  },
};

const LYRICAL_SENTIMENT_DEFAULTS = {
  moody: -0.45,
  melancholic: -0.5,
  sad: -0.6,
  dark: -0.55,
  happy: 0.65,
  upbeat: 0.6,
  euphoric: 0.75,
  chill: 0.15,
  relaxed: 0.1,
  focused: 0.05,
  intense: -0.15,
  aggressive: -0.35,
};

function isMoodSimilar(mood1, mood2) {
  const similar = MOOD_SIMILARITY[mood1] || [];
  return similar.includes(mood2);
}

function isGenreRelated(genre1, genre2) {
  const related = GENRE_RELATIONS[genre1] || [];
  return related.includes(genre2);
}

function resolveDefaults(prefs) {
  const mood = prefs.favorite_mood;
  const acoustic = prefs.likes_acoustic;
  return {
    ...prefs,
    target_popularity:
      prefs.target_popularity ??
      (["happy", "upbeat", "euphoric"].includes(mood) ? 70 : 50),
    target_decade: prefs.target_decade ?? 2010,
    target_instrumentalness:
      prefs.target_instrumentalness ?? (acoustic ? 0.65 : 0.3),
    target_lyrical_sentiment:
      prefs.target_lyrical_sentiment ??
      (LYRICAL_SENTIMENT_DEFAULTS[mood] ?? 0.0),
    target_production_complexity:
      prefs.target_production_complexity ?? (acoustic ? 0.45 : 0.7),
  };
}

export function scoreSong(userPrefs, song, scoringMode = "mood_priority") {
  const prefs = resolveDefaults(userPrefs);
  const weights = SCORING_MODES[scoringMode];
  let score = 0;
  const breakdown = [];

  // 1. Mood match
  if (song.mood === prefs.favorite_mood) {
    score += weights.mood_exact;
    breakdown.push({ label: "Mood match", pts: weights.mood_exact });
  } else if (isMoodSimilar(prefs.favorite_mood, song.mood)) {
    score += weights.mood_similar;
    breakdown.push({ label: "Similar mood", pts: weights.mood_similar });
  }

  // 2. Energy proximity
  const eDiff = Math.abs(song.energy - prefs.target_energy);
  let ePts = 0;
  if (eDiff <= 0.1) ePts = 2.5;
  else if (eDiff <= 0.2) ePts = 1.5;
  else if (eDiff <= 0.3) ePts = 1.0;
  if (ePts) {
    score += ePts;
    breakdown.push({ label: "Energy", pts: ePts });
  }

  // 3. Genre match
  if (song.genre === prefs.favorite_genre) {
    score += weights.genre_exact;
    breakdown.push({ label: "Genre match", pts: weights.genre_exact });
  } else if (isGenreRelated(prefs.favorite_genre, song.genre)) {
    score += weights.genre_related;
    breakdown.push({ label: "Related genre", pts: weights.genre_related });
  }

  // 4. Acousticness
  if (prefs.likes_acoustic) {
    let aPts = 0;
    if (song.acousticness > 0.7) aPts = 1.2;
    else if (song.acousticness > 0.4) aPts = 0.6;
    if (aPts) {
      score += aPts;
      breakdown.push({ label: "Acoustic", pts: aPts });
    }
  }

  // 5. Valence
  let vPts = 0;
  if (song.valence < 0.3) vPts = 0.5;
  else if (song.valence > 0.6) vPts = -0.25;
  if (vPts !== 0) {
    score += vPts;
    breakdown.push({ label: "Valence", pts: vPts });
  }

  // 6. Popularity
  const popDiff = Math.abs(song.popularity - prefs.target_popularity);
  let pPts = 0;
  if (popDiff <= 10) pPts = 1.5;
  else if (popDiff <= 20) pPts = 1.0;
  else if (popDiff <= 30) pPts = 0.5;
  if (pPts) {
    score += pPts;
    breakdown.push({ label: "Popularity", pts: pPts });
  }

  // 7. Decade
  const songDecade = Math.floor(song.release_year / 10) * 10;
  const decadeGap = Math.abs(songDecade - prefs.target_decade);
  let dPts = 0;
  if (decadeGap === 0) dPts = 1.0;
  else if (decadeGap <= 10) dPts = 0.6;
  else if (decadeGap <= 20) dPts = 0.2;
  if (dPts) {
    score += dPts;
    breakdown.push({ label: "Decade", pts: dPts });
  }

  // 8. Instrumentalness
  const instDiff = Math.abs(song.instrumentalness - prefs.target_instrumentalness);
  const instPts = Math.max(0, 1.0 - 2.0 * instDiff);
  if (instPts > 0) {
    score += instPts;
    breakdown.push({ label: "Instrumental", pts: +instPts.toFixed(2) });
  }

  // 9. Lyrical sentiment
  const lyrDiff = Math.abs(song.lyrical_sentiment - prefs.target_lyrical_sentiment);
  const lyrPts = Math.max(0, 1.0 - lyrDiff);
  if (lyrPts > 0) {
    score += lyrPts;
    breakdown.push({ label: "Lyrics", pts: +lyrPts.toFixed(2) });
  }

  // 10. Production complexity
  const prodDiff = Math.abs(
    song.production_complexity - prefs.target_production_complexity
  );
  const prodPts = Math.max(0, 0.8 - 1.6 * prodDiff);
  if (prodPts > 0) {
    score += prodPts;
    breakdown.push({ label: "Production", pts: +prodPts.toFixed(2) });
  }

  return { score: +score.toFixed(2), breakdown };
}

export function recommend(userPrefs, songs, k = 5, scoringMode = "mood_priority") {
  const scored = songs.map((song) => ({
    song,
    ...scoreSong(userPrefs, song, scoringMode),
  }));

  scored.sort((a, b) => b.score - a.score);

  // Artist diversity: max one song per artist
  const seen = new Set();
  const results = [];
  for (const entry of scored) {
    if (results.length >= k) break;
    if (!seen.has(entry.song.artist)) {
      seen.add(entry.song.artist);
      results.push(entry);
    }
  }

  // Backfill if not enough unique artists
  if (results.length < k) {
    for (const entry of scored) {
      if (results.length >= k) break;
      if (!results.includes(entry)) {
        results.push(entry);
      }
    }
  }

  return results;
}

export const GENRES = [
  "alternative rock", "ambient", "classical", "disco", "funk",
  "indie pop", "indie rock", "jazz", "lofi", "metal",
  "pop", "reggae", "rock", "synth-pop", "synthwave",
];

export const MOODS = [
  "aggressive", "chill", "dreamy", "energetic", "euphoric",
  "focused", "happy", "intense", "melancholic", "moody",
  "nostalgic", "relaxed", "upbeat",
];
